Add BigQuery -> ClickHouse SQL reference (#6224)
ClickHouse/clickhouse-docs: docs/cloud/onboard/02_migrate/01_migration_guides/03_bigquery/04_sql-translation-reference.md

<colgroup>
<col />
<col />
</colgroup>
<thead>
<tr>
<th>BigQuery</th>
<th>ClickHouse</th>
</tr>
</thead>
<tbody>

<tr>
<td>

```sql
CREATE TABLE mydataset.t (
  id   INT64,
  name STRING
)
```

</td>
<td>

```sql
CREATE TABLE mydb.t (
  id   Int64,
  name String
)
ENGINE = MergeTree
ORDER BY id
```

</td>
</tr>
<tr>
<td colSpan={2}>

An engine and an `ORDER BY` are required for `MergeTree`-family tables; pick the column(s) that match the query access pattern. See [Sparse primary indexes](/guides/best-practices/sparse-primary-indexes).

</td>
</tr>

<tr>
<td>

```sql
CREATE TABLE IF NOT EXISTS mydataset.t (id INT64)
```

</td>
<td>

```sql
CREATE TABLE IF NOT EXISTS mydb.t (id Int64)
ENGINE = MergeTree
ORDER BY id
```

</td>
</tr>

<tr>
<td>

```sql
CREATE OR REPLACE TABLE mydataset.t (id INT64)
```

</td>
<td>

```sql
CREATE OR REPLACE TABLE mydb.t (id Int64)
ENGINE = MergeTree
ORDER BY id
```

</td>
</tr>

<tr>
<td>

```sql
CREATE TABLE mydataset.t (id INT64)
PARTITION BY DATE(created_at)
```

</td>
<td>

```sql
CREATE TABLE mydb.t (
  id         Int64,
  created_at DateTime
)
ENGINE = MergeTree
PARTITION BY toDate(created_at)
ORDER BY id
```

</td>
</tr>
<tr>
<td colSpan={2}>

> BigQuery partitioning expects a single date/timestamp column; ClickHouse accepts an arbitrary expression. ClickHouse partitions are a storage-organisation feature &mdash; not a substitute for `ORDER BY`.

Reviewer: ```suggestion
BigQuery partitioning expects a single date/timestamp column; ClickHouse accepts an arbitrary expression. ClickHouse partitions are a storage-organization feature &mdash; not a substitute for `ORDER BY`.
```

```

</td>
</tr>

<tr>
<td>

```sql
TRUNCATE TABLE mydataset.t
```

</td>
<td>

```sql
TRUNCATE TABLE mydb.t
```

</td>
</tr>

</tbody>
</table>

### Views and materialized views {#ddl-views}

<table className="sql-translation-table">
<colgroup>
<col />
<col />
</colgroup>
<thead>
<tr>
<th>BigQuery</th>
<th>ClickHouse</th>
</tr>
</thead>
<tbody>

<tr>
<td>

```sql
CREATE VIEW mydataset.v AS
SELECT id, name FROM mydataset.t
```

</td>
<td>

```sql
CREATE VIEW mydb.v AS
SELECT id, name FROM mydb.t
```

</td>
</tr>
<tr>
<td colSpan={2}>

Standard logical views; no storage in either engine.

</td>
</tr>

<tr>
<td>

```sql
CREATE OR REPLACE VIEW mydataset.v AS SELECT ...
```

</td>
<td>

```sql
CREATE OR REPLACE VIEW mydb.v AS SELECT ...
```

</td>
</tr>

<tr>
<td>

```sql
DROP VIEW mydataset.v
```

</td>
<td>

```sql
DROP VIEW mydb.v
```

</td>
</tr>

<tr>
<td>

```sql
CREATE MATERIALIZED VIEW mydataset.mv AS
SELECT status, count(*) AS c
FROM mydataset.t
GROUP BY status
```

</td>
<td>

```sql
CREATE MATERIALIZED VIEW mydb.mv
ENGINE = AggregatingMergeTree
ORDER BY status
AS SELECT status, countState() AS c
FROM mydb.t
GROUP BY status
```

</td>
</tr>
<tr>
<td colSpan={2}>

> ClickHouse [materialized views](/sql-reference/statements/create/view#materialized-view) are incremental insert-time triggers that write into a target table. For an aggregate MV, use [`AggregatingMergeTree`](/engines/table-engines/mergetree-family/aggregatingmergetree) and the `-State` aggregate-function combinator. Query the MV with `countMerge(c)` to finalise.

Reviewer: ```suggestion
ClickHouse [materialized views](/sql-reference/statements/create/view#materialized-view) are incremental insert-time triggers that write into a target table. For an aggregate MV, use [`AggregatingMergeTree`](/engines/table-engines/mergetree-family/aggregatingmergetree) and the `-State` aggregate-function combinator. Query the MV with `countMerge(c)` to finalize.
```

```sql
SELECT element
FROM UNNEST(['a','b','c']) AS element
```

</td>
<td>

```sql
SELECT arrayJoin(['a','b','c']) AS element
```

</td>
</tr>
<tr>
<td colSpan={2}>

Use [`ARRAY JOIN`](/sql-reference/statements/select/array-join) or [`arrayJoin`](/sql-reference/functions/array-join).

</td>
</tr>

<tr>
<td>

```sql
SELECT id FROM mydataset.t
QUALIFY ROW_NUMBER() OVER (
  PARTITION BY user_id
  ORDER BY created_at DESC
) = 1
```

</td>
<td>

```sql
SELECT id FROM (
  SELECT *,
    row_number() OVER (
      PARTITION BY user_id
      ORDER BY created_at DESC
    ) AS rn
  FROM mydb.t
)
WHERE rn = 1
```

</td>
</tr>
<tr>
<td colSpan={2}>

ClickHouse has no `QUALIFY`; wrap the windowed query in a subquery.

</td>
</tr>

<tr>
<td>

```sql
SELECT * FROM mydataset.t TABLESAMPLE SYSTEM (10 PERCENT)
```

</td>
<td>

```sql
SELECT * FROM mydb.t SAMPLE 0.1
```

</td>
</tr>
<tr>
<td colSpan={2}>

ClickHouse [`SAMPLE`](/sql-reference/statements/select/sample) requires the table to declare a `SAMPLE BY` clause.

</td>
</tr>

<tr>
<td>

```sql
SELECT * FROM mydataset.t
FOR SYSTEM_TIME AS OF TIMESTAMP '2024-03-15 00:00:00+00:00'
```

</td>
<td>

No equivalent

</td>
</tr>
<tr>
<td colSpan={2}>

> ClickHouse does not provide row-level time travel. Patterns include [`ReplacingMergeTree`](/engines/table-engines/mergetree-family/replacingmergetree) with a version column or per-day backup tables.

Author: ```suggestion
ClickHouse doesn't provide row-level time travel. Patterns include [`ReplacingMergeTree`](/engines/table-engines/mergetree-family/replacingmergetree) with a version column or per-day backup tables.
```

</td>
</tr>
<tr>
<td colSpan={2}>

ClickHouse `* REPLACE` substitutes the column expression; rename via an outer `SELECT` if you also need the column name to change.

</td>
</tr>

<tr>
<td>

```sql
FROM mydataset.t
|> AGGREGATE sum(amount) AS total
   GROUP BY status
```

</td>
<td>

```sql
SELECT status, sum(amount) AS total
FROM mydb.t
GROUP BY status
```

</td>
</tr>

<tr>
<td>

```sql
FROM mydataset.t
|> ORDER BY created_at DESC
```

</td>
<td>

```sql
SELECT * FROM mydb.t ORDER BY created_at DESC
```

</td>
</tr>

<tr>
<td>

```sql
FROM mydataset.t
|> LIMIT 100
```

</td>
<td>

```sql
SELECT * FROM mydb.t LIMIT 100
```

</td>
</tr>

<tr>
<td>

```sql
FROM mydataset.t
|> JOIN mydataset.u USING (id)
```

</td>
<td>

```sql
SELECT * FROM mydb.t JOIN mydb.u USING (id)
```

</td>
</tr>

<tr>
<td>

```sql
FROM mydataset.t
|> WHERE amount > 100
|> AGGREGATE count(*) GROUP BY status
```

</td>
<td>

```sql
SELECT status, count()
FROM mydb.t
WHERE amount > 100
GROUP BY status
```

</td>
</tr>
<tr>
<td colSpan={2}>

Pipes compose left-to-right; in ClickHouse, layer clauses or use a CTE for readability.

</td>
</tr>

</tbody>
</table>

### Procedural language {#procedural-language}

> ClickHouse SQL is not a procedural language. Variables, loops, statement-level

Author: ```suggestion
ClickHouse SQL isn't a procedural language. Variables, loops, statement-level
```

optionally expand back to rows with
[`arrayJoin`](/sql-reference/functions/array-join). Because of this, many
transformations BigQuery expresses by round-tripping through
[`UNNEST`](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#unnest_operator)
collapse to a single function call in ClickHouse.

<table className="sql-translation-table">
<colgroup>
<col />
<col />
</colgroup>
<thead>
<tr>
<th>BigQuery</th>
<th>ClickHouse</th>
</tr>
</thead>
<tbody>

<tr>
<td>

```sql
ARRAY_CONCAT(a, b)
```

</td>
<td>

```sql
arrayConcat(a, b)
```

</td>
</tr>

<tr>
<td>

```sql
ARRAY_LENGTH(tags)
```

</td>
<td>

```sql
length(tags)
```

</td>
</tr>

<tr>
<td>

```sql
ARRAY_REVERSE(tags)
```

</td>
<td>

```sql
arrayReverse(tags)
```

</td>
</tr>

<tr>
<td>

```sql
ARRAY_TO_STRING(tags, ',')
```

</td>
<td>

```sql
arrayStringConcat(tags, ',')
```

</td>
</tr>

<tr>
<td>

```sql
GENERATE_ARRAY(1, 5)
```

</td>
<td>

```sql
range(1, 6)
```

</td>
</tr>
<tr>
<td colSpan={2}>

ClickHouse [`range`](/sql-reference/functions/array-functions#range) excludes the upper bound.

</td>
</tr>

</tbody>
</table>

### Aggregate functions {#aggregate-functions}

> BigQuery exposes roughly 18 aggregate functions plus a handful of approximate aggregates. ClickHouse ships [more than 150 aggregate functions](/sql-reference/aggregate-functions/reference) and adds [combinators](/sql-reference/aggregate-functions/combinators) — suffixes such as `-If`, `-Array`, `-Map`, `-ForEach`, `-Merge`, and `-State` — that compose with any aggregate to extend its behaviour across data shapes or to use it inside materialized views.

Reviewer: ```suggestion
BigQuery exposes roughly 18 aggregate functions plus a handful of approximate aggregates. ClickHouse ships [more than 150 aggregate functions](/sql-reference/aggregate-functions/reference) and adds [combinators](/sql-reference/aggregate-functions/combinators) — suffixes such as `-If`, `-Array`, `-Map`, `-ForEach`, `-Merge`, and `-State` — that compose with any aggregate to extend its behavior across data shapes or to use it inside materialized views.
```

| [`INTERVAL`](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#interval_type) | No equivalent | Use the [`INTERVAL` expression](/sql-reference/data-types/special-data-types/interval#usage-remarks) or [date/time arithmetic functions](/sql-reference/functions/date-time-functions#addYears) |
| [`JSON`](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#json_type) | [`JSON`](/sql-reference/data-types/newjson) &nbsp;or&nbsp; [`String`](/sql-reference/data-types/string) | `JSON` is preferred; `String` with [`JSONExtract*`](/sql-reference/functions/json-functions) accessors works as a fallback |
| [`NUMERIC`](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#decimal_types) | [`Decimal(P, S)`](/sql-reference/data-types/decimal) | Sized variants `Decimal32(S)` / `Decimal64(S)` / `Decimal128(S)` are also available |
| [`RANGE`](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#range_type) | No equivalent | Store `(start, end)` columns or a `Tuple(start, end)` |

> | [`STRING`](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#string_type) | [`String`](/sql-reference/data-types/string) | Optionally wrap in [`LowCardinality(String)`](/sql-reference/data-types/lowcardinality) for columns with few distinct values (enums, status codes, country codes). String functions are byte-based; the [String family](/sql-reference/functions/string-functions) has `UTF8` variants where relevant |

Author: Feedback from Amy: We should probably flag this as a performance optimization. This might be because I'm in dark mode but I didn't realize those data types were actually links.

| [`STRUCT`](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#constructing_a_struct) | [`Tuple`](/sql-reference/data-types/tuple) &nbsp;or&nbsp; [`Nested`](/sql-reference/data-types/nested-data-structures/nested) | Flattened sibling columns are often more performant in ClickHouse; use `Tuple` / `Nested` for named fields when the nested shape is load-bearing |
| [`TIME`](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#time_type) | No equivalent | Carry a [`DateTime64`](/sql-reference/data-types/datetime64) and extract via [`formatDateTime`](/sql-reference/functions/date-time-functions#formatDateTime) |
| [`TIMESTAMP`](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#timestamp_type) | [`DateTime64(6, 'UTC')`](/sql-reference/data-types/datetime64) | Use for microsecond-precision UTC parity |

## DDL statements {#ddl-statements}

> ### Schemas and databases {#ddl-schemas}

Author: Feedback from Amy: missing the three part namespace callout

<table className="sql-translation-table">
<colgroup>
<col />
<col />
</colgroup>
<thead>
<tr>
<th>BigQuery</th>
<th>ClickHouse</th>
</tr>
</thead>
<tbody>

<tr>
<td>

```sql
CREATE SCHEMA mydataset
```

</td>
<td>

```sql
CREATE DATABASE mydb
```

</td>
</tr>
<tr>
<td colSpan={2}>

BigQuery datasets map to ClickHouse [databases](/sql-reference/statements/create/database).

</td>
</tr>

<tr>
<td>

```sql
CREATE SCHEMA IF NOT EXISTS mydataset
OPTIONS (location = 'US')
```

</td>
<td>

```sql
CREATE DATABASE IF NOT EXISTS mydb
```

</td>
</tr>
<tr>
<td colSpan={2}>

> Location is a service-level decision in ClickHouse Cloud.

Author: Feedback from Amy: decision? maybe configuration?

</td>
</tr>

<tr>
<td>

```sql
ALTER SCHEMA mydataset SET OPTIONS (description = 'sales')
```

</td>
<td>

```sql
ALTER DATABASE mydb MODIFY COMMENT 'sales'
```

</td>
</tr>

<tr>
<td>

```sql
DROP SCHEMA mydataset
```

</td>
<td>

```sql
DROP DATABASE mydb
```

</td>
</tr>
<tr>
<td colSpan={2}>

Add `IF EXISTS` for an idempotent drop in either engine.

</td>
</tr>

<tr>
<td>

```sql
DROP SCHEMA mydataset CASCADE
```

</td>
<td>

```sql
DROP DATABASE mydb
```

</td>
</tr>
<tr>
<td colSpan={2}>

ClickHouse drops tables in the database unconditionally; there is no `CASCADE` keyword.

</td>
</tr>

</tbody>
</table>

### Tables {#ddl-tables}

<table className="sql-translation-table">
<colgroup>
<col />
<col />
</colgroup>
<thead>
<tr>
<th>BigQuery</th>
<th>ClickHouse</th>
</tr>
</thead>
<tbody>

<tr>
<td>

```sql
CREATE TABLE mydataset.t (
  id   INT64,
  name STRING
)
```

</td>
<td>

```sql
CREATE TABLE mydb.t (
  id   Int64,
  name String
)
ENGINE = MergeTree
ORDER BY id
```

</td>
</tr>
<tr>
<td colSpan={2}>

> An engine and an `ORDER BY` are required for `MergeTree`-family tables; pick the columns that match the query access pattern. See [Sparse primary indexes](/guides/best-practices/sparse-primary-indexes).

Author: Feedback from Amy: I think we need a sentence as to why order by is so crucial in clickhouse. it would make the partition comment make more sense. Also link out to what even is a mergetree

</td>
</tr>

</tbody>
</table>

## DCL {#dcl}

### Grants {#dcl-grants}

<table className="sql-translation-table">
<colgroup>
<col />
<col />
</colgroup>
<thead>
<tr>
<th>BigQuery</th>
<th>ClickHouse</th>
</tr>
</thead>
<tbody>

<tr>
<td>

```sql
GRANT SELECT ON TABLE mydataset.t
TO 'user:alice@example.com'
```

</td>
<td>

```sql
GRANT SELECT ON mydb.t TO alice
```

</td>
</tr>
<tr>
<td colSpan={2}>

BigQuery binds to IAM principals (users, groups, service accounts). ClickHouse binds to SQL-level users and roles created with `CREATE USER` / `CREATE ROLE`.

</td>
</tr>

<tr>
<td>

```sql
GRANT ROLE foo ON TABLE mydataset.t
TO 'user:alice@example.com'
```

</td>
<td>

```sql
GRANT foo TO alice
```

</td>
</tr>

<tr>
<td>

```sql
REVOKE SELECT ON TABLE mydataset.t
FROM 'user:alice@example.com'
```

</td>
<td>

```sql
REVOKE SELECT ON mydb.t FROM alice
```

</td>
</tr>

</tbody>
</table>

> ### Roles {#dcl-roles}

Author: Feedback from Amy: I think this table is unnecessary - I think you could just do a sentence and link to the create role/user docs

</td>
</tr>

<tr>
<td>

```sql
SELECT id, sum(amount) OVER w
FROM mydataset.t
WINDOW w AS (PARTITION BY user_id ORDER BY created_at)
```

</td>
<td>

```sql
SELECT id, sum(amount) OVER w
FROM mydb.t
WINDOW w AS (PARTITION BY user_id ORDER BY created_at)
```

</td>
</tr>

<tr>
<td>

```sql
SELECT * FROM mydataset.t
PIVOT (count(*) FOR status IN ('open' AS open, 'closed' AS closed))
```

</td>
<td>

```sql
SELECT
  countIf(status = 'open')   AS open,
  countIf(status = 'closed') AS closed
FROM mydb.t
```

</td>
</tr>
<tr>
<td colSpan={2}>

ClickHouse has no `PIVOT`; use [`countIf` / `sumIf`](/sql-reference/aggregate-functions/combinators#-if) per output column.

</td>
</tr>

<tr>
<td>

```sql
SELECT col, v
FROM (SELECT 1 AS a, 2 AS b)
UNPIVOT (v FOR col IN (a, b))
```

</td>
<td>

```sql
SELECT p.1 AS col, p.2 AS v
FROM (
  SELECT [tuple('a', a), tuple('b', b)] AS pairs FROM mydb.t
)
ARRAY JOIN pairs AS p
```

</td>
</tr>
<tr>
<td colSpan={2}>

ClickHouse has no `UNPIVOT`; emit `(name, value)` tuples and `ARRAY JOIN`.

</td>
</tr>

</tbody>
</table>

> ### Pipe syntax {#pipe-syntax}

Author: Feedback from Amy: you could just say we don't support this?
